src/sys/windows/awakener.rs

```diff
@@ -1,73 +1,39 @@
-use crate::event::Evented;
 use crate::sys::windows::Selector;
-use crate::{io, poll, Interests, PollOpt, Registry, Token};
+use crate::{io, Token};
 use miow::iocp::CompletionStatus;
 use std::sync::Mutex;
 
+#[derive(Debug)]
 pub struct Awakener {
-    inner: Mutex<Option<AwakenerInner>>,
+    inner: Mutex<AwakenerInner>,
 }
 
+#[derive(Debug)]
 struct AwakenerInner {
     token: Token,
     selector: Selector,
 }
 
 impl Awakener {
-    pub fn new() -> io::Result<Awakener> {
+    pub fn new(selector: &Selector, token: Token) -> io::Result<Awakener> {
         Ok(Awakener {
-            inner: Mutex::new(None),
+            inner: Mutex::new(AwakenerInner {
+                selector: selector.clone_ref(),
+                token,
+            }),
         })
     }
 
-    pub fn wakeup(&self) -> io::Result<()> {
+    pub fn wake(&self) -> io::Result<()> {
         // Each wakeup notification has NULL as its `OVERLAPPED` pointer to
         // indicate that it's from this awakener and not part of an I/O
         // operation. This is specially recognized by the selector.
         //
         // If we haven't been registered with an event loop yet just silently
         // succeed.
-        if let Some(inner) = self.inner.lock().unwrap().as_ref() {
-            let status = CompletionStatus::new(0, usize::from(inner.token), 0 as *mut _);
-            inner.selector.port().post(status)?;
-        }
-        Ok(())
-    }
-
-    pub fn cleanup(&self) {
-        // noop
-    }
-}
-
-impl Evented for Awakener {
-    fn register(
-        &self,
-        registry: &Registry,
-        token: Token,
-        events: Interests,
-        opts: PollOpt,
-    ) -> io::Result<()> {
-        assert_eq!(opts, PollOpt::edge());
-        assert_eq!(events, Interests::readable());
-        *self.inner.lock().unwrap() = Some(AwakenerInner {
-            selector: poll::selector(registry).clone_ref(),
-            token: token,
-        });
-        Ok(())
-    }
-
-    fn reregister(
-        &self,
-        registry: &Registry,
-        token: Token,
-        events: Interests,
-        opts: PollOpt,
-    ) -> io::Result<()> {
-        self.register(registry, token, events, opts)
-    }
-
-    fn deregister(&self, _registry: &Registry) -> io::Result<()> {
-        *self.inner.lock().unwrap() = None;
+        let inner = self.inner.lock().unwrap();
+        let status = CompletionStatus::new(0, usize::from(inner.token), 0 as *mut _);
```

Comment:
The windows implementation does not work as one would expect. Receiving completion status events in `select` does not result in a mio event. Instead, in `select`, the callback is called which then results in an event getting posted via the custom readiness queue.

The easiest way to hack in awakeners would probably be to use a hard coded awakener token (there already is one [here](https://github.com/tokio-rs/mio/blob/master/src/sys/windows/selector.rs#L87)). Then, use some other field to pass in the token... the best bet is probably to use the `lpOverlapped` pointer (set to 0 here).

Then, in `select`, when an awakener event is received, push an appropriate event into the events list.

Hopefully we can replace the windows impl soon...

Reply:
Can't we use the `Vec` in `Events`? We detect `overlapped() == 0` in `select` already.

Reply:
I've just pushed https://github.com/tokio-rs/mio/pull/969/commits/37a6ba282aad1fe3ace8d26325f6387eed45f3d6, see if it works.

Reply:
This works, finally.